module/zfs/zvol.c

```diff
@@ -2239,18 +2236,31 @@ zvol_set_snapdev_sync_cb(dsl_pool_t *dp, dsl_dataset_t *ds, void *arg)
 }
 
 /*
- * Traverse all child snapshot datasets and apply snapdev appropriately.
+ * Traverse all child datasets and apply snapdev appropriately.
+ * We call dsl_prop_set_sync_impl() here to set the value only on the toplevel
```

Comment:
> we should set the value for the dataset in question and then we should only be recursing over snapshots if the dataset is a zvol.

This would not deal with the following case:

```
root@debian-8-zfs:~# zfs get snapdev
NAME              PROPERTY  VALUE    SOURCE
testpool          snapdev   hidden   local                   <- zfs set snapdev=visible here
testpool/fs       snapdev   hidden   inherited from testpool
testpool/fs/zvol  snapdev   hidden   inherited from testpool <- snapshot is here
```

Overall, i think we're getting sidetracked here: yes, there can be performance optimizations, but this PR is just about fixing a couple of issues in snapdev inheritance.